### ISSUE TITLE
[R-package] lgb.prepare() and lgb.prepare2() should be simplified

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#2684]
##  Summary

To close this issue, propose a PR that reduces duplicated code in `lgb.prepare_rules()` and `lgb.prepare_rules2()`.

## Motivation

`lgb.prepare_rules()` and `lgb.prepare_rules2()` currently contain a lot of duplicated code and could be simplified. For example, some code is duplicated in the `data.frame` and `data.table` blocks of those functions, such as this snippet

```r
# Get data classes
list_classes <- vapply(data, class, character(1L))

# Map characters/factors
is_fix <- which(list_classes %in% c("character", "factor"))
rules <- list()

# Need to create rules?
if (length(is_fix) > 0L) {
    
    # Go through all characters/factors
    for (i in is_fix) {
        
        # Store column elsewhere
        mini_data <- data[[i]]
```

**Comment by jameslamb** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by jameslamb** [comment]
This issue can be considered permanently closed! Now that #3095 has been merged, there is only one such function (`lgb.convert()`), so this issue is now irrelevant.
